### DISCUSSION TITLE
What type of solver would you use for this problem?

### DISCUSSION THREAD

**Discussion by muhlbach** [cvxpy/cvxpy#1681] [Q&A] [answered]
Hi there! First, I must say I enjoy **cvxpy** so much! That the only library I use for optimization in Python.

Second, the question:

Say I have an (_N_ x _r_)-dim matrix _Y_ and an (_N_ x _d_)-dim matrix _X_. 

I want to find _theta_, an (_d_ x _d_)-dim decision variable that minimizes ***norm2(Y - X @ theta @ beta)***, where @ is the matrix multiplication operator, and beta is some fixed _d_-dim vector for conformity. 

In my case, _N_ is large (~4000), _r_ is small (3), and d is large (1000).

I have tried cp.SCS, but it takes forever (actually it never finished). I can't get neither cp.OSQP nor cp.ECOS to work.

I would love to hear your thoughts on this!

**Comment by SteveDiamond** [comment]
Thanks @muhlbach for the kind words! It might be helpful in your case to introduce a new variable ``Z = theta @ beta``. (So you have ``norm2(Y - X @ Z)``.) I am not 100% sure but it might speed things up.

**Comment by muhlbach** [comment]
@SteveDiamond it me who's thankful!
Yes, I've thought about that too. The thing is, in my problem, ```X @ theta``` has an "interpretation" and is needed further down the road. Essentially, I need ```X``` and ```X @ theta``` to share dimensions, thus ```theta``` has to have _d_ columns (and of course _d_ rows).

I'm definitely open for any suggestions. I can also install other solvers if that might solve the problem?

**Comment by rileyjmurray** [comment, marked as answer]
@muhlbach your current code snippet uses the 2-norm of a matrix-valued expression, which is equal to the largest singular value of the matrix in question and is also known as the spectral norm. Is there any chance you meant to work with (or could work with) the Frobenius norm? If so, then there is a "closed form" solution to your problem. Namely, ``theta = pinv(X) @ Y @ pinv(beta)``, where ``pinv`` is the pseudo-inverse operator. (Ideally, you don't compute the pseudo-inverse explicitly, but it is valid to do that and it's as easy as calling ``np.linalg.pinv``).

**Comment by muhlbach** [comment]
@rileyjmurray Thanks a lot for your suggestion! I was using the spectral norm since it has a nice intuition in my problem, but I'm willing to give that up for a solution that works. Thank you for pointing me to the Frobenius norm, this is working like a charm. Can you elaborate a little bit on why you "ideally" don't compute the pseudo-inverse?

**Reply by muhlbach** [reply]
@rileyjmurray A follow-up question: Having a closer look at your suggestion, I'm wondering why you propose ```theta = pinv(X) @ Y @ pinv(beta)```.

If we define the loss function to be the Frobenius norm, ```L = ||Y - X @ theta @ beta ||_F```, then I would argue that the derivative wrt. ```theta``` would be ```-X'*(Y - X*theta*beta)*beta'```, which would finally lead to ```theta = inv(X'*X)*X'*Y*beta'*pinv(beta*beta'), having replaced ```@```by ```*``` for convenience. What am I missing?

(My sincere apologies if I did the calculus wrong!)

**Reply by rileyjmurray** [reply]
@muhlbach I'm not sure what's meant by "E" in your expression for the (candidate) derivative w.r.t. theta or where "X" goes in your expression for "theta". However, I can say that expressions like ``inv(E'*E)*E'`` arise in pseudo-inverse identities; see [this (part of) this Wikipedia page](https://en.wikipedia.org/wiki/Moore%E2%80%93Penrose_inverse#Definition).

If you want to check that your derivative calculation is correct, the thing to do here is to express ``L`` as the *squared* Frobenius norm. Then you can use ``|| M ||_F^2 = trace(M'M)`` and employ properties like cyclic permuation of trace ``trace(ABC) = trace(CAB) = trace(BCA)`` and linearity of trace. You might find The Matrix Cookbook useful [link to PDF](https://www.math.uwaterloo.ca/~hwolkowi/matrixcookbook.pdf).

**Reply by rileyjmurray** [reply]
We ideally don't compute the pseudo-inverse because it's wasteful in several respects. One usually gets their hands on the pseudo-inverse of ``A`` implicitly by ``U, s, V' = svd(A)`` and then ``pinv_A = V diag(1/s) U'``. If you literally evaluate ``pinv_A`` then you have to multiply matrices of the same sizes as V and U, and then you can apply the pseudo-inverse ``pinv_A @ B``. The alternative is for you to draw parenthesis where you want when computing ``V diag(1/s) U' B``, and that can have big savings if ``B`` is small relative to ``A``.

**Reply by muhlbach** [reply]
> @muhlbach I'm not sure what's meant by "E" in your expression for the (candidate) derivative w.r.t. theta or where "X" goes in your expression for "theta". However, I can say that expressions like `inv(E'*E)*E'` arise in pseudo-inverse identities; see [this (part of) this Wikipedia page](https://en.wikipedia.org/wiki/Moore%E2%80%93Penrose_inverse#Definition).
> 
> If you want to check that your derivative calculation is correct, the thing to do here is to express `L` as the _squared_ Frobenius norm. Then you can use `|| M ||_F^2 = trace(M'M)` and employ properties like cyclic permuation of trace `trace(ABC) = trace(CAB) = trace(BCA)` and linearity of trace. You might find The Matrix Cookbook useful [link to PDF](https://www.math.uwaterloo.ca/~hwolkowi/matrixcookbook.pdf).

@rileyjmurray My apologies. I use a different notation in my paper (using ```E``` instead of ```X```). I have updated the comment to reflect that. To sum up, from minimizing ```L``` wrt. ```theta```, I believe we end up with:
```theta = inv(X'*X)*X'*Y*beta'*pinv(beta*beta')```. Maybe this is not too far from your suggestion with ```theta = pinv(X)*Y*pinv(beta)```.

**Reply by rileyjmurray** [reply]
@muhlbach no worries. I think our expressions are indeed equivalent. You should try to derive the identities
```
pinv(X) = inv(X'*X)*X'
pinv(beta) = beta'*pinv(beta*beta')
```
or show that those relationships might not hold. Deriving pseudo-inverse identities involving an individual matrix is easy. Just express the matrix by its SVD. For example, if ``X`` is a tall matrix with compact SVD `X = U S V'`` (where ``S`` is diagonal), then
```
X'X = (U S V')' (U S V')
    = (V')' (S)' (U)' (U S V')
    = V S U' U S V'
    = V S^2 V'
```
so
```
inv(X'X) X' = inv(V S^2 V') (U S V')'
            = (V S^{-2} V') ((V')' S (U)')
            = V S^{-2} V' V S U'
            = V S^{-2} S U'
            = V S^{-1} U'
            = pinv(X)
```
as desired :). I guess all that's left is showing the second identity!

**Reply by muhlbach** [reply]
Wonderful, thanks @rileyjmurray, this makes a lot of sense and it is working like a charm!!

I was wondering if I could ask you for advice on a related problem? It seems like you're quite the expert on convex optimization! The problem is similar in spirit but slightly more different, and I'm having no luck so far using cvxpy, because it breaks down (forcing the kernel to restart).
